code/modules/power/apc/apc_attack.dm

```diff
@@ -12,68 +14,100 @@
-	var/alert_timer_duration = 0.75 SECONDS
-
-	if(!istype(maybe_stomach, /obj/item/organ/stomach/ethereal))
+	if(isnull(cell))
+		return
+	if(used_stomach.drain_time > world.time)
 		return
-	var/charge_limit = ETHEREAL_CHARGE_DANGEROUS - ETHEREAL_APC_POWER_GAIN
-	var/obj/item/organ/stomach/ethereal/stomach = maybe_stomach
-	var/obj/item/stock_parts/power_store/stomach_cell = stomach.cell
-	if(!((stomach?.drain_time < world.time) && LAZYACCESS(modifiers, RIGHT_CLICK)))
+	if(user.combat_mode)
+		discharge_to_ethereal(user, used_stomach)
+	else
+		charge_from_ethereal(user, used_stomach)
+
+/// Handles discharging our internal cell to an ethereal and their stomach
+/obj/machinery/power/apc/proc/discharge_to_ethereal(mob/living/carbon/human/user, obj/item/organ/stomach/ethereal/used_stomach)
+	var/half_max_charge = cell.max_charge() / 2
+	// Ethereals can't drain APCs under half charge, so that they are forced to look to alternative power sources if the station is running low
+	if(cell.charge() < half_max_charge)
+		addtimer(CALLBACK(src, TYPE_PROC_REF(/atom, balloon_alert), user, "safeties prevent draining!"), ETHEREAL_APC_ALERT_DELAY)
 		return
-	if(ethereal.combat_mode)
-		if(cell.charge <= (cell.maxcharge / 2)) // ethereals can't drain APCs under half charge, this is so that they are forced to look to alternative power sources if the station is running low
-			addtimer(CALLBACK(src, TYPE_PROC_REF(/atom, balloon_alert), ethereal, "safeties prevent draining!"), alert_timer_duration)
+
+	var/obj/item/stock_parts/power_store/stomach_cell = used_stomach.cell
+	used_stomach.drain_time = world.time + ETHEREAL_APC_DRAIN_TIME
+	addtimer(CALLBACK(src, TYPE_PROC_REF(/atom, balloon_alert), user, "draining power"), ETHEREAL_APC_ALERT_DELAY)
+	while(do_after(user, ETHEREAL_APC_DRAIN_TIME, target = src))
+		if(isnull(used_stomach) || (used_stomach != user.get_organ_slot(ORGAN_SLOT_STOMACH)))
+			balloon_alert(user, "stomach removed!?")
 			return
-		if(stomach_cell.charge() > charge_limit)
-			addtimer(CALLBACK(src, TYPE_PROC_REF(/atom, balloon_alert), ethereal, "charge is full!"), alert_timer_duration)
+		if(isnull(cell))
+			balloon_alert(user, "cell removed!")
+			return
+		if(cell.charge() < half_max_charge)
+			balloon_alert(user, "safeties kicked in!")
+			return
+
+		var/our_available_charge = cell.charge() - half_max_charge
+		var/stomach_used_charge = stomach_cell.used_charge()
+		var/potential_charge = min(our_available_charge, stomach_used_charge)
+		var/to_drain = min(ETHEREAL_APC_POWER_GAIN, potential_charge)
+		var/energy_drained = cell.use(to_drain, force = TRUE)
+		used_stomach.adjust_charge(energy_drained)
+
+		if(stomach_cell.used_charge() <= 0)
+			balloon_alert(user, "your charge is full!")
+			return
+		if(cell.charge() <= 0)
+			to_chat(user, span_warning("APC is empty!"))
```

Comment:
Is there a reason the APC full/empty alerts are `to_chat` instead of `balloon_alert`?
Just seems a little inconsistent.

Reply:
because I had a skill issue and forgot to update these